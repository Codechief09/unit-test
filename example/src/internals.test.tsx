import React from 'react';
import addons from '@storybook/addons';
import { render, screen } from '@testing-library/react';

import { composeStories, composeStory } from '../../dist/index';

import * as stories from './components/Button.stories';
import * as globalConfig from '../.storybook/preview';

const { StoryWithParamsAndDecorator } = composeStories(stories);

test('returns composed parameters from story', () => {
  expect(StoryWithParamsAndDecorator.args).toEqual(stories.StoryWithParamsAndDecorator.args);
  expect(StoryWithParamsAndDecorator.parameters).toEqual({
    ...stories.StoryWithParamsAndDecorator.parameters,
    ...globalConfig.parameters,
    component: stories.default.component
  });
  expect(StoryWithParamsAndDecorator.decorators).toEqual([
    ...stories.StoryWithParamsAndDecorator.decorators!,
    ...globalConfig.decorators,
  ]);
});

// common in addons that need to communicate between manager and preview
test('should pass with decorators that need addons channel', () => {
  const PrimaryWithChannels = composeStory(stories.Primary, stories.default, {
    decorators: [
      (StoryFn: any) => {
        addons.getChannel();
        return <StoryFn />;
      },
    ],
  });
  render(<PrimaryWithChannels>Hello world</PrimaryWithChannels>);
  const buttonElement = screen.getByText(/Hello world/i);
  expect(buttonElement).not.toBeNull();
});

describe('Unsupported formats', () => {
  test('should throw error StoryFn.story notation', () => {
    const UnsupportedStory = () => <div>hello world</div>;
    UnsupportedStory.story = { parameters: {} }

    const UnsupportedStoryModule: any = {
      default: {},
      UnsupportedStory
    }

    expect(() => {
      composeStories(UnsupportedStoryModule)
    }).toThrow();
  });

  test('should throw error with non component stories', () => {
    const UnsupportedStoryModule: any = {
      default: {},
      UnsupportedStory: 123
    }

    expect(() => {
      composeStories(UnsupportedStoryModule)
    }).toThrow();
  });
})

describe('non-story exports', () => {
  test('should filter non-story exports with excludeStories', () => {
    const StoryModuleWithNonStoryExports = {
      default: {
        excludeStories: /.*Data/
      },
      LegitimateStory: () => <div>hello world</div>,
      mockData: {}
    }

    const result = composeStories(StoryModuleWithNonStoryExports)
    expect(Object.keys(result)).not.toContain('mockData');
  });

  test('should filter non-story exports with includeStories', () => {
    const StoryModuleWithNonStoryExports = {
      default: {
        includeStories: /.*Story/
      },
      LegitimateStory: () => <div>hello world</div>,
      mockData: {},
    }

    const result = composeStories(StoryModuleWithNonStoryExports)
    expect(Object.keys(result)).not.toContain('mockData');
  });
})

// Batch snapshot testing
const testCases = Object.values(composeStories(stories)).map((Story) => [
  // The ! is necessary in Typescript only, as the property is part of a partial type
  Story.storyName!,
  Story,
])
test.each(testCases)('Renders %s story', async (_storyName, Story) => {
  const tree = await render(<Story />)
  expect(tree.baseElement).toMatchSnapshot()
})