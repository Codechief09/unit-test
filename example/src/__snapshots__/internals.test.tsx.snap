// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders CSF3Button story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    >
      foo
    </button>
  </div>
</body>
`;

exports[`Renders CSF3ButtonWithRender story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <div>
      <p
        data-testid="custom-render"
      >
        I am a custom render function
      </p>
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders InputFieldFilled story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <input />
  </div>
</body>
`;

exports[`Renders Primary story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <button
      class="storybook-button storybook-button--large storybook-button--primary"
      type="button"
    >
      foo
    </button>
  </div>
</body>
`;

exports[`Renders Secondary story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    >
      Children coming from story args!
    </button>
  </div>
</body>
`;

exports[`Renders StoryWithDecoratorUseHook story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <div>
      <div>
        args.size:
        large
      </div>
      <button
        class="storybook-button storybook-button--large storybook-button--primary"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders StoryWithParamsAndDecorator story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    >
      foo
    </button>
  </div>
</body>
`;

exports[`Renders WithLocale story 1`] = `
<body>
  <div>
    <div>
      Locale: 
      en
    </div>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    >
      Hello!
    </button>
  </div>
</body>
`;
